### PR TITLE
HACK week: Migration of client/utils/account-fees to TS

### PR DESCRIPTION
**Pull request by mgascam**
#### Changes proposed in this Pull Request

TypeScript migration for `client/utils/account-fees`

#### Testing instructions

* Tests should complete without errors.

-------------------

- [x] Added changelog entries to both `readme.txt` and `changelog.txt` (or does not apply)
- [x] Covered with tests (or have a good reason not to test in description ☝️)
- [x] Tested on mobile (or does not apply)

**Post merge**

- [x] Added testing instructions to the [release testing doc](https://github.com/Automattic/woocommerce-payments/wiki/Release-testing-instructions) (or does not apply)

### CODE REVIEW

File: client/utils/account-fees.tsx
Comment (kalessil) [inline client/utils/account-fees.tsx:102]: Looks fishy, are the definitions correct?

File: client/utils/account-fees.tsx
Comment (kalessil) [inline client/utils/account-fees.tsx:106]: Looks fishy, are the definitions correct?

File: client/utils/account-fees.tsx
Comment (kalessil) [inline client/utils/account-fees.tsx:110]: Return type is missing

File: client/utils/account-fees.tsx
Comment (kalessil) [inline client/utils/account-fees.tsx:248]: Looks not good, why not keep the original construct?

Comment (mgascam) [reply client/utils/account-fees.tsx:248]: @kalessil the TS compiler did not like the original construct, it was complaining that it could be `undefined`. I'll try to find a solution

File: client/utils/account-fees.tsx
Comment (mgascam) [reply client/utils/account-fees.tsx:102]: thanks @kalessil  for checking it. As far as I understand the code, the `additional` and `fx` are optional properties so I marked them like that in the type definition. For that reason the TS compiler was complaining that those values could be `undefined`, that's way I fallback to 0.

I'll double check type definitions as you suggest.

File: client/utils/account-fees.tsx
Comment (mgascam) [reply client/utils/account-fees.tsx:106]: Same as previous comment, I'll double check definitions

File: client/utils/account-fees.tsx
Comment (kalessil) [reply client/utils/account-fees.tsx:102]: The fields are presented, but might have null as a value.

Speaking in definitions terms, that would be:
```php
export interface DiscountFee {
	end_time: string | null;
	volume_allowance: number | null;
	volume_currency: string | null;
	current_volume: number | null;
	percentage_rate: number;
	fixed_rate: number;
	discount: number|null;
}
```

To re-check the definition, please follow the trail in `\WCPay\Service\Fee_Service::get_account_fee_info` on the server side.

File: client/utils/account-fees.tsx
Comment (kalessil) [reply client/utils/account-fees.tsx:248]: Hm, then the re is an incomplete definition somewhere on the way there.

File: client/utils/account-fees.tsx
Comment (mgascam) [reply client/utils/account-fees.tsx:102]: Thanks @kalessil this is key point, I was not realizing that the properties were there but could have null values

File: client/utils/account-fees.tsx
Comment (mgascam) [reply client/utils/account-fees.tsx:86]: I believe it is not possible to specify return type only as `DiscountFee`. Since this function can return `BaseFee` or `DiscountFee`, and `BaseFee` does not have all the properties that `DiscountFee` has. Please let me know if maybe I am missing something 😅

File: client/utils/account-fees.tsx
Comment (mgascam) [reply client/utils/account-fees.tsx:247]: I could not find a different way of doing this.. The way I see it: `currentFee` can be of two different types `BaseFee` or `DiscountFee`. Since `BaseFee` does not have the `discount` property, the compiler complaints. 

I am afraid that I reached a dead end here...To be honest, I believe it is fair to check if the property exists in this way (it is covered in TS docs [here](https://www.typescriptlang.org/docs/handbook/2/narrowing.html#the-in-operator-narrowing) but please let me know if there is a better way.

File: client/utils/test/account-fees.tsx
Comment (mgascam) [inline client/utils/test/account-fees.tsx:185]: After introduction of `BaseFee` and `DiscountFee` types, this test is no longer possible.

**Review by mordeth — APPROVED** [review]
LGTM!

File: client/components/account-status/account-fees/account-fees.d.ts
Comment (anu-rock) [inline client/components/account-status/account-fees/account-fees.d.ts:1]: I think we finalized `client/types` as the location to store all .d.ts files.

File: client/utils/account-fees.tsx
Comment (anu-rock) [reply client/utils/account-fees.tsx:247]: At first, I had the same reaction as @kalessil. But I think I see the point now.

File: client/utils/declarations.d.ts
Comment (anu-rock) [inline client/utils/declarations.d.ts:1]: This file may be moved to `client/types/utils`.

File: client/components/account-status/account-fees/account-fees.d.ts
Comment (kalessil) [reply client/components/account-status/account-fees/account-fees.d.ts]: Fixed in 1e25ce1e4cd7dd41fa1a9ef22e996614a682b9e7

File: client/utils/declarations.d.ts
Comment (kalessil) [reply client/utils/declarations.d.ts]: This declaration is not needed. Addressed in be848db1f8aa4e9a7d5abad5e7ae6ca0fb9c662e.

File: client/utils/account-fees.tsx
Comment (kalessil) [reply client/utils/account-fees.tsx:82]: Corect!

File: client/utils/account-fees.tsx
Comment (kalessil) [reply client/utils/account-fees.tsx]: Addressed in c17fb6880af211def0bbe8d6def9bff350ec514c

File: client/utils/test/account-fees.tsx
Comment (kalessil) [reply client/utils/test/account-fees.tsx:185]: Restored the test, see 7c4f1bc67373183d73a68114414e54a1f6dd6eb9

**Review by kalessil — APPROVED** [review]
LGTM! @mgascam  did a great job and some challenges were indeed non-trivial.

I did a few commits to address the issues blocking the PR and we are ready to merge now.
@mgascam, please do one more manual testing round before merging this to ensure the fees rendering works on account overview and transactions details pages (timeline).

**Comment by mgascam** [comment]
> LGTM! @mgascam did a great job and some challenges were indeed non-trivial.
> 
> I did a few commits to address the issues blocking the PR and we are ready to merge now. @mgascam, please do one more manual testing round before merging this to ensure the fees rendering works on account overview and transactions details pages (timeline).

Thanks for your contributions @kalessil! I did some manual tests on account overview and transaction details timeline and everything works as expected. I am proceeding to merge.

**Pull request merged**
